File: solution/264. Ugly Number II/264_03.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        static vector<int> uglys{1};
        static int i2 = 0, i3 = 0, i5 = 0;
        while (uglys.size() < n) {
            const int next2 = uglys[i2]*2;
            const int next3 = uglys[i3]*3;
            const int next5 = uglys[i5]*5;
            const int next = min(next2, min(next3, next5));
            uglys.push_back(next);
            if (next == next2) i2++;
            if (next == next3) i3++;
            if (next == next5) i5++;
        }
        return uglys[n-1];
    }
};

File: solution/264. Ugly Number II/264_03.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        res = [1]
        i2 = i3 = i5 = 0
        for _ in range(n):
            next2 = res[i2] * 2
            next3 = res[i3] * 3
            next5 = res[i5] * 5
            nxt = min(next2, next3, next5)
            res.append(nxt)
            if nxt == next2:
                i2 += 1
            if nxt == next3:
                i3 += 1
            if nxt == next5:
                i5 += 1
        return res[n - 1]
    
